wit/streams.wit

```diff
@@ -259,4 +259,40 @@ interface streams {
             len: u64,
         ) -> result<u64, stream-error>;
     }
+
+    /// Continuously read all data from the input stream and write it to the
+    /// output stream.
+    ///
+    /// Forwarding is done on a background task. The tasks runs until either
+    /// stream closes or fails. After that, the streams are dropped and the
+    /// returned future is resolved. Dropping the future before it has been
+    /// resolved aborts the background task and drops the streams.
+    ///
+    /// The future resolves with the reason that caused the forwarding to terminate.
+    /// In case one of the streams ended normally, this will be `stream-error::closed`.
+    ///
+    /// This function requires exclusive access to the provided streams,
+    /// yet it does not change the ownership structure of them. I.e. they remain
+    /// children of their parent resources. If the parents place any lifetimes
+    /// restrictions on their children, those continue to apply.
+    forward: func(src: input-stream, dst: output-stream) -> future-forward-result;
+
+    /// Represents a future which may eventually return trailers, or an error.
+    resource future-forward-result {
+        /// Returns a pollable which becomes ready when either the result is
+        /// available. When this pollable is ready, the `get` method will return
+        /// `some`.
+        subscribe: func() -> pollable;
+
+        /// Get the resolved value.
+        ///
+        /// Returns:
+        /// - `none`: when the future hasn't resolved yet. Use `subscribe` to wait
+        ///   for its completion.
+        /// - `some(error(_))`: when the future is resolved, but the result value
+        ///   has already been taken before.
+        /// - `some(ok(value))`: when the future is ready with a value. This is
+        ///   returned only once.
+        get: func() -> option<result<stream-error>>;
```

Comment:
We also need this get function to have a way to represent "youve already retrieved the result of this and i cant give it to you again", in http we used an outer result for that, so it would end up like `option<result<result<stream-error>, _>>`

Reply:
Indeed, that's documented as:

```
- `some(error(_))`: when the future is resolved, but the result value has already been taken before.
```

Reply:
oh, sorry, I read the type signature before the docs here. How do we represent the case where the forward completed with success? there is no stream-error to represent success, you need a `result<stream-error>` for that.

Reply:
Well "success" in this case is when one of the streams ended with `closed`. Which is already captured in the `stream-error` type.

Reply:
OK, but as a user I would want to distinguish between whether the input-stream closed (expected outcome of the operation) and the output-stream closed (unexpected). When the expected outcome happens, we should return an `ok`.

Additionally I'd consider returning a `variant forward-error { source(stream-error), destination(stream-error) }` to distinguish where the error came from - we shouldnt throw away information that the user will need to debug.

Reply:
> OK, but as a user I would want to distinguish between whether the input-stream closed (expected outcome of the operation) and the output-stream closed (unexpected). When the expected outcome happens, we should return an ok.

Fair enough.

> Additionally I'd consider returning a variant forward-error { source(stream-error), destination(stream-error) } to distinguish where the error came from - we shouldnt throw away information that the user will need to debug.

I took this design from `splice` which already tosses errors from both streams on the same pile. Are you sure we want to introduce this inconsistency & complexity in return types? Or is it OK that say that whenever a consumer really cares about which stream failed, they can just attempt `input-stream::read(0)` or `output-stream::check-write()` after the fact?